src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
+        currentNode.prev = node;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        Node<T> currentNode = findNode(index);
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndex(index);
+        Node<T> currentNode = findNode(index);
+        T oldValiue = currentNode.value;
+        currentNode.value = value;
+        return oldValiue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        Node<T> currentNode = findNode(index);
+        unlink(currentNode);
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (currentNode.value == null && object == null
+                    || currentNode != null && currentNode.value.equals(object)) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node == head && size == 1) {
+            head = null;
+            tail = null;
+            size--;
+            return;
+        }
+        if (node == head) {
+            head = node.next;
+            node.next.prev = null;
+            size--;
+            return;
+        }
+        if (node == tail) {
+            tail = node.prev;
+            node.prev.next = null;
+            size--;
+            return;
+        }
+        node.prev.next = node.next;
+        node.next.prev = node.prev;
+        size--;
+    }
+
+    private void validateIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("There is no such index.");
+        }
+    }
+
+    private Node<T> findNode(int index) {
+        int counter = 0;
+        if (index < size / 2) {
+            Node<T> currentNode = head;
+            while (currentNode != null) {
+                if (counter == index) {
+                    break;
+                }
+                counter++;
+                currentNode = currentNode.next;
+            }
+            return currentNode;
+        }
+        index = size - index - 1;
+        Node<T> currentNode = tail;
+        while (currentNode != null) {
+            if (counter == index) {
+                break;
+            }
+            counter++;
+            currentNode = currentNode.prev;
+        }
```

Comment:
I think this logic is too difficult, you can use to for-loops, first from head to tail, second from tail to head